### PR TITLE
FALCON-2077 Api support for Process SLA

### PR DESCRIPTION
**Pull request by PraveenAdlakha**

### CODE REVIEW

**Comment by PraveenAdlakha** [comment]
Result of Dev Testing : 

$ bin/falcon entity -type process -slaAlert -name ProcessInstanceRerunTest-agregator-33 -start 2016-08-30T07:30Z -end 2016-08-30T08:30Z
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/usr/local/falcon/falcon-0.11-SNAPSHOT/client/lib/slf4j-log4j12-1.7.5.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/usr/local/falcon/falcon-0.11-SNAPSHOT/client/lib/falcon-shell-0.11-SNAPSHOT.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
name: ProcessInstanceRerunTest-agregator-33, type: PROCESS, cluster: local, instanceTime: 2016-08-30T07:30Z, tags: Missed-SLA-High
name: ProcessInstanceRerunTest-agregator-33, type: PROCESS, cluster: local, instanceTime: 2016-08-30T07:35Z, tags: Missed-SLA-High
name: ProcessInstanceRerunTest-agregator-33, type: PROCESS, cluster: local, instanceTime: 2016-08-30T07:40Z, tags: Missed-SLA-High
name: ProcessInstanceRerunTest-agregator-33, type: PROCESS, cluster: local, instanceTime: 2016-08-30T07:45Z, tags: Missed-SLA-High
name: ProcessInstanceRerunTest-agregator-33, type: PROCESS, cluster: local, instanceTime: 2016-08-30T07:50Z, tags: Missed-SLA-High
name: ProcessInstanceRerunTest-agregator-33, type: PROCESS, cluster: local, instanceTime: 2016-08-30T07:55Z, tags: Missed-SLA-High
name: ProcessInstanceRerunTest-agregator-33, type: PROCESS, cluster: local, instanceTime: 2016-08-30T08:00Z, tags: Missed-SLA-High
name: ProcessInstanceRerunTest-agregator-33, type: PROCESS, cluster: local, instanceTime: 2016-08-30T08:05Z, tags: Missed-SLA-High
name: ProcessInstanceRerunTest-agregator-33, type: PROCESS, cluster: local, instanceTime: 2016-08-30T08:10Z, tags: Missed-SLA-High
name: ProcessInstanceRerunTest-agregator-33, type: PROCESS, cluster: local, instanceTime: 2016-08-30T08:15Z, tags: Missed-SLA-High
name: ProcessInstanceRerunTest-agregator-33, type: PROCESS, cluster: local, instanceTime: 2016-08-30T08:20Z, tags: Missed-SLA-High
name: ProcessInstanceRerunTest-agregator-33, type: PROCESS, cluster: local, instanceTime: 2016-08-30T08:25Z, tags: Missed-SLA-High
name: ProcessInstanceRerunTest-agregator-33, type: PROCESS, cluster: local, instanceTime: 2016-08-30T08:30Z, tags: Missed-SLA-High

**Comment by PraveenAdlakha** [comment]
Hi All,

EntitySLAAlertServiceTest might fail just wanted to make sure its not my machine issue.Will see what is needed to be done if it fails.

Thanks

File: common/src/main/java/org/apache/falcon/persistence/BeanUtils.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    /*
+    * Checks wheather Entity Type is in feed or process.
+    *
+    * @param  entityType
+    *         The {@code String} to check it against valid {@code enum} EntityType.
+    * @throws FalconException
+    *         if the entityType is not valid.
+    * */
+
+    public static void checkEntityType(String entityType)throws FalconException {
```
Comment (pallavi-rao) [inline common/src/main/java/org/apache/falcon/persistence/BeanUtils.java]: Can't you use EntityType.getEnum(entityType) instead?

File: common/src/main/resources/startup.properties [hunk 2]
```diff
@@ -98,7 +98,7 @@
 #Configurations used in UTs
 debug.config.store.uri=file://${user.dir}/target/store
 #Location to store state of Feed SLA monitoring service
-debug.feed.sla.service.store.uri= file://${user.dir}/target/data/sla/pendingfeedinstances
+debug.entity.sla.service.store.uri= file://${user.dir}/target/data/sla/pendingfeedinstances
```
Comment (pallavi-rao) [inline common/src/main/resources/startup.properties]: The store location should be `file://${user.dir}/target/data/sla/pendingEntityInstances`

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java [hunk 3]
```diff
@@ -48,7 +48,7 @@ public void putMonitoredEntity(String entityName, String entityType) throws Falc
 
         MonitoredEntityBean monitoredEntityBean = new MonitoredEntityBean();
         monitoredEntityBean.setEntityName(entityName);
-        monitoredEntityBean.setEntityType(entityType);
+        monitoredEntityBean.setEntityType(entityType.toLowerCase());
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: Why this change?

File: prism/src/main/java/org/apache/falcon/resource/AbstractSchedulableEntityManager.java [hunk 4]
```diff
@@ -111,16 +111,29 @@ protected synchronized void scheduleInternal(String type, String entity, Boolean
...
+     * @param colo colo in which entity is to be looked into
+     * @throws FalconException if the validation fails
+     * **/
+
     public static void validateSlaParams(String entityType, String entityName, String start, String end,
                                          String colo) throws FalconException {
         EntityType type = EntityType.getEnum(entityType);
-        if (type != EntityType.FEED) {
+        LOG.info("EntityName:" + entityName);
+        if (!Arrays.asList(EntityType.FEED.toString().toLowerCase(), EntityType.PROCESS.toString().toLowerCase()).
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/resource/AbstractSchedulableEntityManager.java]: You can instead use type != EntityType.FEED || type != EntityType.PROCESS

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java (hunk 3 above)
Comment (PraveenAdlakha) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: By Default we were storing in capitals but in case of api call we are passing entityType whatever user provides i.e process,Process or PROCESS in case the db is case sensitive we will not get any results because of the where clause in the SQL.

As entityType is internal to falcon so I thought of storing and querying it in lowercase.Let me know your thoughts on this.

**Comment by pallavi-rao** [comment]
@PraveenAdlakha, are you taking up corresponding CLI changes as part of a separate JIRA?

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java (hunk 3 above)
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: Then, lets change it in the Bean itself as that is closest to the DB and the conversation will need to be done in lesser number of places.

**Comment by pallavi-rao** [comment]
My bad. Didn't notice that you had tested against CLI itself. Which means the CLI changes were already in place. Please ignore my comment on the CLI change.

File: prism/src/main/java/org/apache/falcon/resource/AbstractSchedulableEntityManager.java [hunk 5]
```diff
@@ -111,16 +111,29 @@ protected synchronized void scheduleInternal(String type, String entity, Boolean
         }
     }
 
+    /**
+     * Validates the parameters wheather SLA is supported or not.
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/resource/AbstractSchedulableEntityManager.java]: Nit : Typo. Should be "whether"

File: common/src/main/java/org/apache/falcon/persistence/BeanUtils.java (hunk 1 above)
Comment (PraveenAdlakha) [inline common/src/main/java/org/apache/falcon/persistence/BeanUtils.java]: After discussed moving it to EntityType as EntityType.getEnum will not serve the complete purpose of the function.

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java (hunk 3 above)
Comment (PraveenAdlakha) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: yes will make this change but we still need it in this class because the parameter entityType comes all the way through CLI and we have to convert it to lower case before sending it to DB in the where clause to get the results.

File: prism/src/main/java/org/apache/falcon/resource/AbstractSchedulableEntityManager.java (hunk 4 above)
Comment (PraveenAdlakha) [inline prism/src/main/java/org/apache/falcon/resource/AbstractSchedulableEntityManager.java]: this is the reason I used Arrays.asList it should be 
 type != EntityType.FEED && type != EntityType.PROCESS
...its more readable  :)

Anyway found a better after our discussion type.isScheduleable ...

File: client/src/main/java/org/apache/falcon/entity/v0/EntityType.java [hunk 6]
```diff
@@ -101,6 +101,16 @@ public boolean isSchedulable() {
         return ((this != EntityType.CLUSTER) && (this != EntityType.DATASOURCE));
     }
 
+    public static void assertSchedulable(String entityType){
```
Comment (pallavi-rao) [inline client/src/main/java/org/apache/falcon/entity/v0/EntityType.java]: Can you use EntityType.getEnum inside this method? This method will be pretty much similar to AbstractSchedulableEntityManager.checkSchedulableEntity, except, the exception may be different.

**Comment by pallavi-rao** [comment]
Other than the minor nit. +1.

**Comment by PraveenAdlakha** [comment]
@pallavi-rao have made the changes please review.

**Comment by pallavi-rao** [comment]
Merging this...
